pkg/cvo/metrics.go

```diff
@@ -105,72 +115,171 @@ type asyncResult struct {
+
+	certDir := filepath.Dir(certFile)
+	keyDir := filepath.Dir(keyFile)
+
+	origCertChecksum, err := checksumFile(certFile)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize certificate file checksum: %w", err)
+	}
+	origKeyChecksum, err := checksumFile(keyFile)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize key file checksum: %w", err)
+	}
+
+	// Set up and start the file watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if watcher == nil || err != nil {
+		return fmt.Errorf("Failed to create file watcher for certificate and key rotation: %w", err)
+	} else {
+		defer watcher.Close()
+		if err := watcher.Add(certDir); err != nil {
+			return fmt.Errorf("Failed to add %v to watcher: %w", certDir, err)
+		}
+		if certDir != keyDir {
+			if err := watcher.Add(keyDir); err != nil {
+				return fmt.Errorf("Failed to add %v to watcher: %w", keyDir, err)
+			}
+		}
+	}
 
 	shutdown := false
+	restartServer := false
 	var loopError error
 	for resultChannelCount > 0 {
 		if shutdown {
 			select {
 			case result := <-resultChannel:
 				resultChannelCount--
-				if result.error == nil {
-					klog.Infof("Collected metrics %s goroutine.", result.name)
-				} else {
-					klog.Errorf("Collected metrics %s goroutine: %v", result.name, result.error)
-					loopError = result.error
-				}
+				loopError = handleServerResult(result, loopError)
 			case <-shutdownContext.Done(): // out of time
 				klog.Errorf("Abandoning %d uncollected metrics goroutines", resultChannelCount)
 				return shutdownContext.Err()
 			}
 		} else {
 			select {
 			case <-runContext.Done(): // clean shutdown
-			case result := <-resultChannel: // crashed before a shutdown was requested
+			case result := <-resultChannel: // crashed before a shutdown was requested or metrics server recreated
+				if restartServer {
+					klog.Info("Creating metrics server with updated TLS configuration.")
+					server = createHttpServer()
+					go startListening(server, tlsConfig, listenAddress, resultChannel)
+					restartServer = false
+					continue
+				}
 				resultChannelCount--
-				if result.error == nil {
-					klog.Infof("Collected metrics %s goroutine.", result.name)
+				loopError = handleServerResult(result, loopError)
+			case event := <-watcher.Events:
+				if event.Op != fsnotify.Chmod && event.Op != fsnotify.Remove {
+					if changed, err := certsChanged(origCertChecksum, origKeyChecksum, certFile, keyFile); changed {
+
+						// Update file checksums with latest files.
+						//
+						if origCertChecksum, err = checksumFile(certFile); err != nil {
+							klog.Errorf("Failed to update certificate file checksum: %v", err)
+							loopError = err
+						}
+						if origKeyChecksum, err = checksumFile(keyFile); err != nil {
+							klog.Errorf("Failed to update key file checksum: %v", err)
+							loopError = err
+						}
+
+						tlsConfig, err = makeTLSConfig(certFile, keyFile)
+						if err == nil {
```

Comment:
if we failed either of the `checksumFile` calls above, the `makeTLSConfig` call here will have clobbered `err`, and we might have a successful `makeTLSConfig` call get us through to this restart, when really we want to fall through to shutdown.  Maybe `if loopError == nil`?

Reply:
Oh crap, yeah. Let me re-structure so it just bails as soon as it it's any of those errors rather than continuing. That was unintentional.

Reply:
I just added a `break` to each one.

Reply:
I don't think `break` works, because the shutdown code is currently inside the `for` loop, and after `Shutdown`, we also have the wait on `resultChannel` and/or a timeout in this same `for` loop, so we can consider that error when deciding which error to complain about when the function returns.  I think dropping the `break` and using `if loopError == nil` here will work.  "If we haven't hit an error yet, try to restart, otherwise, continue to shutdown".

Reply:
These `break`s break out of the select not the `for` loop. So it falls through to line 279 and goes through the normal `shutdown` logic.